metal/src/kernels/matmul/ggml_gemm/mod.rs

```diff
@@ -33,14 +34,14 @@ impl GemmKernel for GgmlGemm {
         "ggml"
     }
 
-    fn is_supported_dts(&self, dts: &[DatumType]) -> TractResult<bool> {
-        ensure!(dts.len() == 2);
+    fn is_supported_dts(&self, facts: &[TypedFact]) -> bool {
+        assert!(facts.len() == 2, "Ggml: Expected 2 inputs for Matmul");
 
-        if dts[0] == DatumType::F32 {
-            Ok(dts[1] == DatumType::F32)
-        } else {
-            Ok(dts[0] == DatumType::F16 && matches!(dts[1], DatumType::F32 | DatumType::F16))
-        }
+        (as_q40_fact(&facts[1]).is_some()
+            && matches!(facts[0].datum_type, DatumType::F16 | DatumType::F32))
+            || ((facts[1].datum_type == DatumType::F32) && (facts[0].datum_type == DatumType::F32))
+            || ((facts[1].datum_type == DatumType::F16)
+                && matches!(facts[0].datum_type, DatumType::F32 | DatumType::F16))
```

Comment:
lol, this is very hard to parse :) maybe we could split the tricky q40 case apart then do somethings like
```
use DatumType::*;
match (facts[0].datum_type, facts[1].datum_type) {
  (F32, F32) => true,
  (F16, F16) => true,
  (F16, F32) => true,
  _ => false
}
```